test_ioccc/man/man8/mkiocccentry_slots.8
.\" section 8 man page for mkiocccentry_slots.sh
.\"
.\" This man page was first written by Landon Curt Noll for the IOCCC
.\" in 2025.
.\"
.\" "Be careful what you measure, because you are likely to get it."
.\"				    -- Richard Feynman to Landon Curt Noll
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH mkiocccentry_slots.sh 8 "29 August 2025" "mkiocccentry_slots.sh" "IOCCC tools"
.SH NAME
.B mkiocccentry_slots.sh
\- test chkentry on good and bad submission directories
.SH SYNOPSIS
.B mkiocccentry_slots.sh
.RB [\| \-h \|]
.RB [\| \-V \|]
.RB [\| \-v
.IR level \|]
.RB [\| \-g
.IR good_tree \|]
.RB [\| \-M
.IR mkiocccentry \|]
.RB [\| \-Z
.IR repo_topdir \|]
.SH DESCRIPTION
.B mkiocccentry_slots.sh
runs
.BR mkiocccentry (1)
against slot data found under
.I good_tree.
.PP
This tool looks for
.B UUID-SLOT_NUM
information, where
.B UUID
is a version 4 UUID of the form:
.sp
.RS
xxxxxxxx-xxxx-4xxx-Nxxx-xxxxxxxxxxxx
.RE
.sp
where
.I x
is a hex character,
.I 4
is the UUID version, and
.I N
ne of 8, 9, a, or b.
.PP
A
.I SLOT_NUM
is an IOCCC slot number, currently a single decimal digit.
.PP
The following in an example of a
.BR UUID-SLOT_NUM :
.sp
.RS
12345678-1234-4321-abcd-1234567890ab-6
.RE
.sp
.PP
Each
.B UUID-SLOT_NUM
must have both an answers file:
.BR good_tree/answers/UUID-SLOT_NUM ,
and a code directory:
.BR good_tree/topdir/UUID-SLOT_NUM .
.PP
When
.BR mkiocccentry (1)
successfully processes a slat, a submission tarball is formed:
.BR good_tree/workdir/UUID-SLOT_NUM.txz ,
and a submission tree directory:
.BR good_tree/workdir/UUID-SLOT_NUM/ .
.PP
If
.BR mkiocccentry (1)
successfully processes all slots, then
.B mkiocccentry_slots.sh
will exit 0, otherwise it will exit non-zero.
.SH OPTIONS
.TP
.B \-h
Print help and exit.
.TP
.B \-V
Print version and exit.
.TP
.BI \-v\  level
Set verbosity level to
.IR level
(def: 0).
.TP
.BI \-M\  mkiocccentry
Set path to mkiocccentry to
.IR mkiocccentry .
By default, the path:
.B ./mkiocccentry
(relative to repo_topdir) is used.
.TP
.BI \-g\  good_tree
Set path to the good slot tree.
By default, the path:
.B test_ioccc/slot/good
(relative to repo_topdir) is used.
.TP
.BI \-Z\  repo_topdir
Set the top level repo directory for the mkiocccentry toolkit.
By default, the current directory or the parent if the current directory is used.
.SH EXIT STATUS
.TP
0
all OK
.TQ
1
at least one test failed.
.TQ
2
.B \-h
and help string printed or
.B \-V
and version string printed
.TQ
3
invalid command line
.TQ
4
missing slot information
.TQ
>=10
internal error
.SH FILES
Yes.  :-)
.SH BUGS
If you have a problem with the tool (not JSON itself! :\-) ) you can report it at the GitHub issues page.
You can open an issue at:
.sp
.RS 4
.ft B
https://github.com/ioccc\-src/mkiocccentry/issues
.ft R
.RE
.SH SEE ALSO
.BR mkiocccentry (1),
.BR txzchk (1),
.BR chksubmit (1).


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 2.6.2 2025-09-04

Make adjustments to the new `test_file_util.c` so that it will function in cases
where someone has a file called `foobar` in their directory (this would be me as
I use it as part of testing mkiocccentry).

A man page was started but not added yet because the exit codes in the test code
need to be fixed first and that cannot happen now.

Updated `FILE_UTIL_TEST_VERSION` to `"2.0.1 2025-09-04"`.


## Release 2.6.1 2025-09-02

Moved nearly all of the tree and file utility functions
that were in `jparse/util.c` and `jparse/util.h` over to the
files `soup/file_util.c` and `soup/file_util.h`.
The code that was moved was not needed for the JSON
parser nor the parser's related utilities.
Moving that code will both simplify the JSON
parser code base and make it easier to improve and
enhance tree and file utility functions.
Only a few small utility functions such as `exists()`,
`is_file()`, and `is_read()` remain behind as private
functions in `jparse.l` and `jparse/test_jparse/jnum_gen.c`.

Moved the `jparse/test_jparse/util_test` code over to
as `test_ioccc/test_file_util.c` and ran via the
`test_ioccc/ioccc_test.sh` test script.  That test script
is in its own file and thus no longer has to be copied
from some other source file in order to be compiled.

Restored some of the double blank lines separating functions
in `jparse/util.c`.

The `jparse/test_jparse/run_jparse_tests.sh` no longer runs `util_test`.
Changed the `jparse/test_jparse/jnum_chk` test exit code to `27` to be in sequence.

Made `.gitignore` and `jparse/.gitignore` much more robust, ignoring
intermediate files that may be created while performing `make release`.

Removed `UTIL_TEST` from the `jparse/` tree.

Removed `UTIL_TEST_VERSION` from the `jparse/` tree..  That symbol
has become, in `soup/version.h`, the symbol `FILE_UTIL_TEST_VERSION`
and now has the value "2.0.0 2025-09-01".

Updated, improved and sorted `.gitignore` and `jparse/.gitignore`.

Updated `RUN_JPARSE_TESTS_VERSION` to "2.0.1 2025-09-01".

Updated `JPARSE_UTILS_VERSION` to "2.1.0 2025-09-01".

Updated `JPARSE_REPO_VERSION` to "2.3.0 2025-09-01".

**NO** JSON parser code was changed as a result of
these moves.  Nevertheless the movement of code
serves as a "_checkpoint_" tag and thus warrants
a 2nd-level version number change.


## Release 2.6.0 2025-09-01

Updated `MKIOCCCENTRY_REPO_VERSION` to "2.6.0 2025-09-01"
in anticipation of work that may be needed for **IOCCC29**,
and to tag this version as the starting code base for
**IOCCC29** related development.

Updated `MIN_MKIOCCCENTRY_VERSION` to `MKIOCCCENTRY_VERSION`
which is currently "2.1.2 2025-08-30".

Improved `txzchk_test(8)` with warnings about changing `-d txzchk_tree`.

Fixed echo of `test_ioccc/ioccc_test.sh` command line within
the `test_ioccc/ioccc_test.sh` script.

Synced [jparse repo](https://github.com/xexyl/jparse) and fixed make depend.


## Release 2.5.2 2025-08-30

Fix bugs in txzchk and mkiocccentry.

Although it mostly matters for `islower()` and `isupper()` it is safer to check
`isascii()` before passing it on to one of the `is*()` functions. This is
because of how some implementations do the functions. This probably does not
matter much but it's now safer.

Be more aggressive in errors with txzchk. Although it's still a warning for most
issues with the tarball if a NULL pointer is encountered rather than skipping to
the next line it is an error.

Fix assumption about `find_utils()` in mkiocccentry and txzchk. Before freeing
the `char *` of each tool, also check that it was found. The calls to
`find_utils()` do (for the tools needed) do pass non-NULL and if the tool is
already specified at the command line it still strdup()s it. But the two tools
now check `found_foo` before freeing the strings.

Updated `MKIOCCCENTRY_VERSION` to `"2.1.2 2025-08-30"`.
Updated `TXZCHK_VERSION` to `"2.0.6 2025-08-30"`.

Fixed debug typo in `test_ioccc/chksubmit_test.sh` about a test that should fail.

Changed `test_ioccc/chksubmit_test.sh` IOCCC_TEST_VERSION from "2.1.0 2025-08-28"
to "2.1.1 2025-08-31".

Synced with the [dbg repo](https://github.com/lcn2/dbg), including new git tags.

Synced with the [dyn_array repo](https://github.com/lcn2/dyn_array), including new git tags.

Fixed dependencies in `jparse/test_jparse/Makefile`.

Synced `jparse/version.h` with the [jparse repo](https://github.com/xexyl/jparse) `version.h`.

Synced `jparse/CHANGES.md` with the [jparse repo](https://github.com/xexyl/jparse) `CHANGES.md`.

Updated jparse/CHANGES.md` as needed.


## Release 2.5.1 2025-08-28

Add new tool `chksubmit(1)` as part of issue #1299. This tool runs `chkentry -S`
(new option, see below) and `mkiocccentry(1)` runs it directly instead of
`chkentry`. The `-C` option to `mkiocccentry(1)` now specifies the path to the
new `chksubmit(1)` tool.

Two new options were added to `chkentry(1)` but neither are implemented because
the details are not yet finalised. This is `-S` for `submission mode` and `-s`
for special mode.

The new tool has a man page, `chksubmit.1`.

The template JSON files had major updates due to the update of the `info` and
`auth` structs. Thus the version of these also have been updated.

The test `txzchk` files had to be updated as well due to call to the updated
`find_utils()` function.

Fixed display bug in `mkiocccentry(1)` (inconsistency with a value checked and
what is reported).

Updated the `forbidden_filenames` array. It is not clear if this is complete
yet.

Updated `MKIOCCCENTRY_VERSION` to `"2.1.1 2025-08-28"`.
Updated `INFO_VERSION` to `"1.18 2025-05-28"`.
Updated `AUTH_VERSION` to `"1.24 2025-08-28"`.
Updated `TXZCHK_VERSION` to `"2.0.6 2025-08-28"`.
Updated `CHKENTRY_VERSION` to `"2.1.0 2025-08-28"`.
Updated `gen_test_JSON.sh` version to `"1.0.4 2025-08-28"`.
Set `CHKSUBMIT_VERSION` to `"1.0.0 2025-08-28"`.

The `chksubmit(1)` tool now passes `-S` to `chkentry(1)`.


test_ioccc/mkiocccentry_slots.sh
#!/usr/bin/env bash
#
# mkiocccentry_slots.sh - rebuild the slot test_ioccc/slot/good/workdir directories using mkiocccentry
#
# Copyright (c) 2025 by Landon Curt Noll.  All Rights Reserved.
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby granted,
# provided that the above copyright, this permission notice and text
# this comment, and the disclaimer below appear in all of the following:
#
#       supporting documentation
#       source copies
#       source works derived from this source
#       binaries derived from this source or from derived source
#
# CODY BOONE FERGUSON DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT
# SHALL CODY BOONE FERGUSON BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
# DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
# TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
# OF THIS SOFTWARE.
#
# Share and enjoy! :-)
#     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
#

# setup


# IOCCC requires use of C locale
#
export LANG="C"
export LC_CTYPE="C"
export LC_NUMERIC="C"
export LC_TIME="C"
export LC_COLLATE="C"
export LC_MONETARY="C"
export LC_MESSAGES="C"
export LC_PAPER="C"
export LC_NAME="C"
export LC_ADDRESS="C"
export LC_TELEPHONE="C"
export LC_MEASUREMENT="C"
export LC_IDENTIFICATION="C"
export LC_ALL="C"


# setup
#
export MKIOCCCENTRY_SLOTS_VERSION="1.0.0 2025-08-28"
export MKIOCCCENTRY="./mkiocccentry"
export GOOD_TREE="test_ioccc/slot/good"
#
export EXIT_CODE=0
export REPO_TOPDIR=

# usage message
#
export USAGE="usage: $0 [-h] [-V] [-v level] [-g good_tree] [-M mkiocccentry] [-Z repo_topdir] [-k] [-B]

    -h			print help and exit
    -V			print version and exit
    -v level		set verbosity level for this script: (def level: 0)
    -M mkiocccentry	path to mkiocccentry executable (def: $MKIOCCCENTRY)
    -g good_tree	path to the good slot tree (def: $GOOD_TREE)
    -Z repo_topdir	top level repo directory for the mkiocccentry toolkit (def: try . or ..)

Exit codes:
     0   all OK
     1   at least one test failed
     2   -h and help string printed or -V and version string printed
     3   invalid command line
     4	 missing slot information
 >= 10   internal error or missing file or directory

mkiocccentry_slots.sh version: $MKIOCCCENTRY_SLOTS_VERSION"


# parse args
#
export V_FLAG="0"
while getopts :hVv:M:g:Z: flag; do
    case "$flag" in
    h)	echo "$USAGE" 1>&2
	exit 2
	;;
    V)	echo "$MKIOCCCENTRY_SLOTS_VERSION"
	exit 2
	;;
    v)	V_FLAG="$OPTARG";
	;;
    M)	MKIOCCCENTRY="$OPTARG";
	;;
    g)	GOOD_TREE="$OPTARG";
	;;
    Z)  REPO_TOPDIR="$OPTARG";
        ;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
	echo 1>&2
	echo "$USAGE" 1>&2
	exit 3
	;;
    :)	echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
	echo 1>&2
	echo "$USAGE" 1>&2
	exit 3
	;;
   *)
	;;
    esac
done

# check args
#
shift $(( OPTIND - 1 ));
if [[ $# -ne 0 ]]; then
    echo "$0: ERROR: expected 0 arguments, found $#" 1>&2
    echo 1>&2
    echo "$USAGE" 1>&2
    exit 3
fi

# change to the top level directory as needed
#
if [[ -n $REPO_TOPDIR ]]; then
    if [[ ! -d $REPO_TOPDIR ]]; then
	echo "$0: ERROR: -Z $REPO_TOPDIR given: not a directory: $REPO_TOPDIR" 1>&2
	exit 3
    fi
    if [[ $V_FLAG -ge 1 ]]; then
	echo "$0: debug[1]: -Z $REPO_TOPDIR given, about to cd $REPO_TOPDIR" 1>&2
    fi
    # SC2164 (warning): Use 'cd ... || exit' or 'cd ... || return' in case cd fails.
    # https://www.shellcheck.net/wiki/SC2164
    # shellcheck disable=SC2164
    cd "$REPO_TOPDIR"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: -Z $REPO_TOPDIR given: cd $REPO_TOPDIR exit code: $status" 1>&2
	exit 3
    fi
elif [[ -f mkiocccentry.c ]]; then
    REPO_TOPDIR="$PWD"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: assume REPO_TOPDIR is .: $REPO_TOPDIR" 1>&2
    fi
elif [[ -f ../mkiocccentry.c ]]; then
    cd ..
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: cd .. exit code: $status" 1>&2
	exit 3
    fi
    REPO_TOPDIR="$PWD"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: assume REPO_TOPDIR is ..: $REPO_TOPDIR" 1>&2
    fi
else
    echo "$0: ERROR: cannot determine REPO_TOPDIR, use -Z topdir" 1>&2
    exit 3
fi
if [[ $V_FLAG -ge 3 ]]; then
    echo "$0: debug[3]: REPO_TOPDIR is the current directory: $REPO_TOPDIR" 1>&2
fi

# check for mkiocccentry
#
if [[ ! -e $MKIOCCCENTRY ]]; then
    echo "$0: ERROR: mkiocccentry not found: $MKIOCCCENTRY" 1>&2
    exit 10
fi
if [[ ! -f $MKIOCCCENTRY ]]; then
    echo "$0: ERROR: mkiocccentry not a regular file: $MKIOCCCENTRY" 1>&2
    exit 11
fi
if [[ ! -x $MKIOCCCENTRY ]]; then
    echo "$0: ERROR: mkiocccentry not executable: $MKIOCCCENTRY" 1>&2
    exit 12
fi

# check that the good slot tree is a writable directory
#
if [[ ! -e $GOOD_TREE ]]; then
    echo "$0: ERROR: good slot tree not found: $GOOD_TREE" 1>&2
    exit 13
fi
if [[ ! -d $GOOD_TREE ]]; then
    echo "$0: ERROR: good slot tree not a directory: $GOOD_TREE" 1>&2
    exit 14
fi
if [[ ! -w $GOOD_TREE ]]; then
    echo "$0: ERROR: good slot tree not writable directory: $GOOD_TREE" 1>&2
    exit 15
fi

# verify good slot tree topdir sub-directory is writable
#
export GOOD_TOPDIR="$GOOD_TREE/topdir"
if [[ ! -e $GOOD_TOPDIR ]]; then
    echo "$0: ERROR: good slot tree topdir sub-directory directory not found: $GOOD_TOPDIR" 1>&2
    exit 16
fi
if [[ ! -d $GOOD_TOPDIR ]]; then
    echo "$0: ERROR: good slot tree topdir sub-directory not a directory: $GOOD_TOPDIR" 1>&2
    exit 17
fi
if [[ ! -w $GOOD_TOPDIR ]]; then
    echo "$0: ERROR: good slot tree topdir sub-directory not writable directory: $GOOD_TOPDIR" 1>&2
    exit 18
fi

# verify good slot tree answers sub-directory is writable
#
export GOOD_ANSWERS="$GOOD_TREE/answers"
if [[ ! -e $GOOD_ANSWERS ]]; then
    echo "$0: ERROR: good slot tree answers sub-directory directory not found: $GOOD_ANSWERS" 1>&2
    exit 19
fi
if [[ ! -d $GOOD_ANSWERS ]]; then
    echo "$0: ERROR: good slot tree answers sub-directory not a directory: $GOOD_ANSWERS" 1>&2
    exit 20
fi
if [[ ! -w $GOOD_ANSWERS ]]; then
    echo "$0: ERROR: good slot tree answers sub-directory not writable directory: $GOOD_ANSWERS" 1>&2
    exit 21
fi
export GOOD_BUILD_OUT="$GOOD_TREE/build.out"
export GOOD_WORKDIR="$GOOD_TREE/workdir"

# UUID regex
#
# A UUID has the 36 character format:
#
#     xxxxxxxx-xxxx-4xxx-Nxxx-xxxxxxxxxxxx
#
# where 'x' is a hex character, 4 is the UUID version, and N is one of 8, 9, a, or b.
#
export RE_HEX='[0-9A-Fa-f]'			    # 1 hex
export RE_HEX_3="$RE_HEX$RE_HEX$RE_HEX"		    # 3 hex
export RE_HEX_4="$RE_HEX_3$RE_HEX"		    # 4 hex
export RE_N='[89ABab]'				    # N is one of 8, 9, a, or b
export RE_UUID_1="$RE_HEX$RE_HEX$RE_HEX$RE_HEX"	    # 4 hex
export RE_UUID_0="$RE_UUID_1$RE_UUID_1"		    # 8 hex
export RE_UUID_2="4$RE_HEX$RE_HEX$RE_HEX"	    # UUID version 4
export RE_UUID_3="$RE_N$RE_HEX_3"		    # N + 3 hex
export RE_UUID_4="$RE_HEX_4$RE_UUID_0"		    # 12 hex
# UUID regex for IOCCC
export RE_UUID="$RE_UUID_0-$RE_UUID_1-$RE_UUID_2-$RE_UUID_3-$RE_UUID_4"

# slot number regex
#
export RE_SLOT_NUM='[0-9]'			    # IOCCC slot number is 1 digit

# UUID-SLOT_NUM regex
#
export RE_UUID_SLOT_NUM="$RE_UUID-$RE_SLOT_NUM"	    # IOCCC UUID-SLOT_NUM

# form temporary exit code file
#
TMP_EXIT_CODE=$(mktemp -u .exit_code.XXXXXXXXXX)
export TMP_EXIT_CODE
trap 'rm -f $TMP_EXIT_CODE; exit' 0 1 2 3 15
rm -f "$TMP_EXIT_CODE"
echo "0" > "$TMP_EXIT_CODE"
if [[ ! -s "$TMP_EXIT_CODE" ]]; then
    echo "$0: ERROR: could not create exit code file: $TMP_EXIT_CODE" 1>&2
    exit 22
fi
if [[ ! -w "$TMP_EXIT_CODE" ]]; then
    echo "$0: ERROR: exit code file not writable: $TMP_EXIT_CODE" 1>&2
    exit 23
fi

# debugging
#
if [[ $V_FLAG -ge 3 ]]; then
    echo "$0: debug[3]: MKIOCCCENTRY_SLOTS_VERSION=$MKIOCCCENTRY_SLOTS_VERSION" 1>&2
    echo "$0: debug[3]: MKIOCCCENTRY=$MKIOCCCENTRY" 1>&2
    echo "$0: debug[3]: GOOD_TREE=$GOOD_TREE" 1>&2
    echo "$0: debug[3]: EXIT_CODE=$EXIT_CODE" 1>&2
    echo "$0: debug[3]: REPO_TOPDIR=$REPO_TOPDIR" 1>&2
    echo "$0: debug[3]: GOOD_TOPDIR=$GOOD_TOPDIR" 1>&2
    echo "$0: debug[3]: GOOD_ANSWERS=$GOOD_ANSWERS" 1>&2
    echo "$0: debug[3]: GOOD_BUILD_OUT=$GOOD_BUILD_OUT" 1>&2
    echo "$0: debug[3]: GOOD_WORKDIR=$GOOD_WORKDIR" 1>&2
    echo "$0: debug[3]: RE_UUID=$RE_UUID" 1>&2
    echo "$0: debug[3]: RE_SLOT_NUM=$RE_SLOT_NUM" 1>&2
    echo "$0: debug[3]: RE_UUID_SLOT_NUM=$RE_UUID_SLOT_NUM" 1>&2
    echo "$0: debug[3]: TMP_EXIT_CODE=$TMP_EXIT_CODE" 1>&2
fi

# form an empty build.out sub-directory of good slot tree
#
if [[ $V_FLAG -ge 3 ]]; then
    echo "$0: debug[3]: about to form an empty build.out sub-directory of good slot tree: $GOOD_BUILD_OUT" 1>&2
fi
rm -rf "$GOOD_BUILD_OUT"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: rm -rf $GOOD_BUILD_OUT failed, exit code: $status" 1>&2
    exit 24
fi
mkdir -p -- "$GOOD_BUILD_OUT"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: mkdir -p -- $GOOD_BUILD_OUT failed, exit code: $status" 1>&2
    exit 25
fi

# form an empty workdir sub-directory of good slot tree
#
if [[ $V_FLAG -ge 3 ]]; then
    echo "$0: debug[3]: about to form an empty workdir sub-directory of good slot tree: $GOOD_WORKDIR" 1>&2
fi
rm -rf "$GOOD_WORKDIR"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: rm -rf $GOOD_WORKDIR failed, exit code: $status" 1>&2
    exit 26
fi
mkdir -p -- "$GOOD_WORKDIR"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: mkdir -p -- $GOOD_WORKDIR failed, exit code: $status" 1>&2
    exit 27
fi

# process all slots
#
# We look for UUID-SLOT_NUM's that have both:
#
#   answers file:   $GOOD_ANSWERS/UUID-SLOT_NUM
#   code directory: $GOOD_TOPDIR/UUID-SLOT_NUM/
#
# Then we use mkiocccentry to generate a:
#
#   submission: $GOOD_WORKDIR/UUID-SLOT_NUM/12345678-1234-4321-abcd-1234567890ab-0/
#
# The submission is a directory that contains both:
#
#   submission tarball:	UUID-SLOT_NUM.txz
#   submission tree:	UUID-SLOT_NUM/
#
# The submission tree, if mkiocccentry is successful, will have:
#
#   non-empty auth JSON:	UUID-SLOT_NUM/.auth.json
#   non-empty info JSON:	UUID-SLOT_NUM/.info.json
#   prog.c:			UUID-SLOT_NUM/prog.c
#   non-empty Makefile:		UUID-SLOT_NUM/Makefile
#   non-empty remarks.md:	UUID-SLOT_NUM/remarks.md
#
# This code will attempt to process ALL UUID-SLOT_NUM's.
# All UUID-SLOT_NUM's are successful, the EXIT_CODE will be set to 0.
# If any UUID-SLOT_NUM fails, the EXIT_CODE will be set to non-zero.
#
find "$GOOD_ANSWERS" -type f -name '[0-9a-f]*-[0-9a-f]' -print |
    sed -e 's;^.*/;;' |
    grep -E "^$RE_UUID_SLOT_NUM$" |
    LANG=C sort |
    while read -r UUID_SLOT_NUM; do

	# verify that the answers file for UUID-SLOT_NUM is a readable non-empty file
	#
	if [[ ! -s $GOOD_ANSWERS/$UUID_SLOT_NUM || ! -r $GOOD_ANSWERS/$UUID_SLOT_NUM ]]; then
	    echo "$0: ERROR: not a readable non-empty answers file: $GOOD_ANSWERS/$UUID_SLOT_NUM" 1>&2
	    TMP_EXIT=1
	    echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
	    echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
	    continue
	fi

	# verify we have a topdir for this UUID-SLOT_NUM
	#
	if [[ -d $GOOD_TOPDIR/$UUID_SLOT_NUM ]]; then

	    # remove any existing workdir
	    #
	    if [[ -d $GOOD_WORKDIR/$UUID_SLOT_NUM ]]; then
		if [[ $V_FLAG -ge 3 ]]; then
		    echo "$0: debug[3]:" rm -rf "${GOOD_WORKDIR/$UUID_SLOT_NUM:?}" 1>&2
		fi
		rm -rf "${GOOD_WORKDIR/$UUID_SLOT_NUM:?}"
		status="$?"
		if [[ $status -ne 0 ]]; then
		    echo "$0: ERROR: rm -rf $GOOD_WORKDIR/$UUID_SLOT_NUM failed, error code: $status" 1>&2
		    TMP_EXIT=1
		    echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
		    echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
		    continue
		fi
		if [[ -d $GOOD_WORKDIR/$UUID_SLOT_NUM ]]; then
		    echo "$0: ERROR: failed to remove existing workdir: $GOOD_WORKDIR/$UUID_SLOT_NUM" 1>&2
		    TMP_EXIT=1
		    echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
		    echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
		    continue
		fi
	    fi

	    # create a new empty workdir
	    #
	    if [[ $V_FLAG -ge 3 ]]; then
		echo "$0: debug[3]:" mkdir -p -- "$GOOD_WORKDIR/$UUID_SLOT_NUM" 1>&2
	    fi
	    mkdir -p -- "$GOOD_WORKDIR/$UUID_SLOT_NUM"
	    status="$?"
	    if [[ $status -ne 0 ]]; then
		TMP_EXIT=1
		echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
		echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
		continue
	    fi

	    # remove any build.out file
	    #
	    if [[ -e $GOOD_BUILD_OUT/$UUID_SLOT_NUM.out ]]; then
		if [[ $V_FLAG -ge 3 ]]; then
		    echo "$0: debug[3]:" rm -f "$GOOD_BUILD_OUT/$UUID_SLOT_NUM.out" 1>&2
		fi
		rm -f "$GOOD_BUILD_OUT/$UUID_SLOT_NUM.out"
		status="$?"
		if [[ $status -ne 0 ]]; then
		    echo "$0: ERROR: rm -f $GOOD_BUILD_OUT/$UUID_SLOT_NUM.out failed, error code: $status" 1>&2
		    TMP_EXIT=1
		    echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
		    echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
		continue
		    continue
		fi
		if [[ -e $GOOD_BUILD_OUT/$UUID_SLOT_NUM.out ]]; then
		    echo "$0: ERROR: failed to remove existing build.out: $GOOD_BUILD_OUT/$UUID_SLOT_NUM.out" 1>&2
		    TMP_EXIT=1
		    echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
		    echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
		    continue
		fi
	    fi

	    # run mkiocccentry
	    #
	    if [[ $V_FLAG -ge 1 ]]; then
		echo "$0: debug[1]:" "$MKIOCCCENTRY" -Y -i "$GOOD_ANSWERS/$UUID_SLOT_NUM" \
			"$GOOD_WORKDIR/$UUID_SLOT_NUM" "$GOOD_TOPDIR/$UUID_SLOT_NUM" \
			"> $GOOD_BUILD_OUT/$UUID_SLOT_NUM.out 2>&1" 1>&2
	    fi
	    "$MKIOCCCENTRY" -Y -i "$GOOD_ANSWERS/$UUID_SLOT_NUM" \
		    "$GOOD_WORKDIR/$UUID_SLOT_NUM" "$GOOD_TOPDIR/$UUID_SLOT_NUM" \
		    > "$GOOD_BUILD_OUT/$UUID_SLOT_NUM.out" 2>&1
	    status="$?"
	    if [[ $status -ne 0 ]]; then
		echo "$0: ERROR: mkiocccentry on $UUID_SLOT_NUM failed, error code: $status" 1>&2
		TMP_EXIT=1
		echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
		echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
		continue
	    fi

	# case: topdir for this UUID-SLOT_NUM is missing
	#
	else
	    echo "$0: ERROR: missing good slot tree topdir sub-directory: $GOOD_TOPDIR/$UUID_SLOT_NUM" 1>&2
	    TMP_EXIT=1
	    echo "$TMP_EXIT" > "$TMP_EXIT_CODE"
	    echo "$0: Warning: set exit code: $TMP_EXIT" 1>&2
	fi
    done

# All Done!!! All Done!!! -- Jessica Noll, Age 2
#
if [[ -s "$TMP_EXIT_CODE" ]]; then
    EXIT_CODE=$(< "$TMP_EXIT_CODE")
fi
if [[ $V_FLAG -ge 1 ]]; then
    if [[ $EXIT_CODE -eq 0 ]]; then
	echo "$0: debug[1]: all tests PASSED" 1>&2
    else
	echo "$0: debug[1]: some tests FAILED" 1>&2
    fi
fi
if [[ $V_FLAG -ge 1 && $EXIT_CODE -gt 0 ]]; then
    echo "$0: ERROR: about to exit: $EXIT_CODE" 1>&2
fi
exit "$EXIT_CODE"
